File: libs/ui-canvas/src/main/kotlin/mono/html/canvas/mouse/MouseEventObserver.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.html.canvas.mouse

import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.html.canvas.canvas.DrawingInfoController
import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import mono.livedata.MutableLiveData
import mono.livedata.distinctUntilChange
import mono.ui.appstate.state.ScrollMode
import org.w3c.dom.HTMLDivElement
import org.w3c.dom.events.MouseEvent
import org.w3c.dom.events.WheelEvent
import kotlin.js.Date

/**
 * A class to attach and observe mouse events happening on the canvas.
 * All events will be combined into a single live data [mousePointerLiveData].
 */
internal class MouseEventObserver(
    lifecycleOwner: LifecycleOwner,
    container: HTMLDivElement,
    drawingInfoLiveData: LiveData<DrawingInfoController.DrawingInfo>,
    shiftKeyStateLiveData: LiveData<Boolean>,
    private val scrollModeLiveData: LiveData<ScrollMode>
) {
    private val mousePointerMutableLiveData = MutableLiveData<MousePointer>(MousePointer.Idle)
    val mousePointerLiveData: LiveData<MousePointer> =
        mousePointerMutableLiveData.distinctUntilChange()

    private var drawingInfo: DrawingInfoController.DrawingInfo =
        drawingInfoLiveData.value

    private var mouseWheelDeltaX: Float = 0F
    private var mouseWheelDeltaY: Float = 0F
    private val drawingOffsetPointPxMutableLiveData = MutableLiveData(Point.ZERO)
    val drawingOffsetPointPxLiveData: LiveData<Point> = drawingOffsetPointPxMutableLiveData

    private val mouseDoubleClickDetector = MouseDoubleClickDetector()

    init {
        container.onmousedown = ::setMouseDownPointer
        container.onmouseup = ::setMouseUpPointer
        container.onmousemove = ::setMouseMovePointer
        container.onwheel = ::setMouseWheel

        drawingInfoLiveData.observe(lifecycleOwner) {
            drawingInfo = it
        }
        shiftKeyStateLiveData.observe(lifecycleOwner) {
            val currentMouseValue = mousePointerLiveData.value
            if (currentMouseValue is MousePointer.Drag) {
                mousePointerMutableLiveData.value = currentMouseValue.copy(isWithShiftKey = it)
            }
        }
    }

    fun forceUpdateOffset(offsetPx: Point) {
        drawingOffsetPointPxMutableLiveData.value = offsetPx
    }

    private fun setMouseDownPointer(event: MouseEvent) {
        if (mousePointerLiveData.value == MousePointer.Idle ||
            mousePointerLiveData.value is MousePointer.Move
        ) {
            mousePointerMutableLiveData.value =
                MousePointer.Down(event.toPoint(), event.toPointPx(), event.shiftKey)
        }
    }

    private fun setMouseUpPointer(event: MouseEvent) {
        val isDoubleClick = mouseDoubleClickDetector.onMouseUp()
        val currentValue = mousePointerLiveData.value
        val clickPoint = event.toPoint()

        mousePointerMutableLiveData.value = when (currentValue) {
            is MousePointer.Down ->
                MousePointer.Up(currentValue.point, clickPoint, event.shiftKey)

            is MousePointer.Drag ->
                MousePointer.Up(currentValue.mouseDownPoint, clickPoint, event.shiftKey)

            is MousePointer.Move,
            is MousePointer.Up,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            MousePointer.Idle -> MousePointer.Idle
        }

        if (currentValue is MousePointer.Down) {
            mousePointerMutableLiveData.value = if (isDoubleClick) {
                MousePointer.DoubleClick(event.toPoint())
            } else {
                MousePointer.Click(clickPoint, event.shiftKey)
            }
        }
        mousePointerMutableLiveData.value = MousePointer.Idle
    }

    private fun setMouseMovePointer(event: MouseEvent) {
        mouseDoubleClickDetector.reset()
        val newPointer = when (val mousePointer = mousePointerLiveData.value) {
            is MousePointer.Down ->
                MousePointer.Drag(mousePointer.point, event.toPoint(), event.shiftKey)
                    .takeIf { it.point != mousePointer.point }

            is MousePointer.Drag ->
                MousePointer.Drag(mousePointer.mouseDownPoint, event.toPoint(), event.shiftKey)

            is MousePointer.Move,
            is MousePointer.Up,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            MousePointer.Idle -> MousePointer.Move(event.toPoint(), event.toPointPx())
        } ?: return
        mousePointerMutableLiveData.value = newPointer
    }

    private fun setMouseWheel(event: WheelEvent) {
        event.preventDefault()
        event.stopPropagation()

        val scrollHorizontalThresholdPx = SCROLL_THRESHOLD_PIXEL
        val scrollVerticalThresholdPx = SCROLL_THRESHOLD_PIXEL

        val (scrollDeltaX, scrollDeltaY) = event.getScrollDelta()

        val wheelDeltaLeft = scrollDeltaX * SCROLL_SPEED_RATIO / scrollHorizontalThresholdPx
        val wheelDeltaTop = scrollDeltaY * SCROLL_SPEED_RATIO / scrollVerticalThresholdPx
        val accumulateWheelDeltaLeft = mouseWheelDeltaX + wheelDeltaLeft
        val accumulateWheelDeltaTop = mouseWheelDeltaY + wheelDeltaTop

        val usableDeltaLeft = accumulateWheelDeltaLeft.toInt()
        val usableDeltaTop = accumulateWheelDeltaTop.toInt()

        if (usableDeltaLeft != 0 || usableDeltaTop != 0) {
            val offsetLeft =
                drawingInfo.offsetPx.left - usableDeltaLeft * scrollHorizontalThresholdPx.toInt()
            val offsetTop =
                drawingInfo.offsetPx.top - usableDeltaTop * scrollVerticalThresholdPx.toInt()
            drawingOffsetPointPxMutableLiveData.value = Point(offsetLeft, offsetTop)
        }

        mouseWheelDeltaX = accumulateWheelDeltaLeft - usableDeltaLeft
        mouseWheelDeltaY = accumulateWheelDeltaTop - usableDeltaTop
    }

    /**
     * Returns scroll delta x and y of the wheel event after adjusting with meta keys:
     * - When Alt/Option key is pressed, swap delta x and delta y (this is for those who use mouse)
     * - When Shift key is pressed, scroll mode other than [ScrollMode.BOTH] will be flipped
     *   (vertical -> horizontal, horizontal -> vertical)
     * - Returns the delta x and delta y of the wheel event according to adjusted scroll mode (with
     *   Shift key) and delta value (with Alt/Option key):
     *   - Vertical: delta x = 0
     *   - Horizontal: delta y = 0
     *   - Both: delta x and delta y = delta value
     */
    private fun WheelEvent.getScrollDelta(): Pair<Float, Float> {
        val deltaX = deltaX.toFloat()
        val deltaY = deltaY.toFloat()
        val scrollDeltaX = if (altKey) deltaY else deltaX
        val scrollDeltaY = if (altKey) deltaX else deltaY

        val scrollMode = when (scrollModeLiveData.value) {
            ScrollMode.BOTH -> ScrollMode.BOTH
            ScrollMode.VERTICAL -> if (shiftKey) ScrollMode.HORIZONTAL else ScrollMode.VERTICAL
            ScrollMode.HORIZONTAL -> if (shiftKey) ScrollMode.VERTICAL else ScrollMode.HORIZONTAL
        }
        return when (scrollMode) {
            ScrollMode.BOTH -> scrollDeltaX to scrollDeltaY
            ScrollMode.VERTICAL -> 0.0F to scrollDeltaY
            ScrollMode.HORIZONTAL -> scrollDeltaX to 0.0F
        }
    }

    private fun MouseEvent.toPoint(): Point =
        Point(
            drawingInfo.toBoardColumn(offsetX.toInt()),
            drawingInfo.toBoardRow(offsetY.toInt())
        )

    private fun MouseEvent.toPointPx(): Point = Point(offsetX.toInt(), offsetY.toInt())

    /**
     * A class for detecting double click.
     * If two mouse ups happen within 500ms, it's a double click.
     */
    private class MouseDoubleClickDetector {
        private var lastMouseUpMillis: Long = 0
        private var mouseUpCount: Int = 0

        fun onMouseUp(): Boolean {
            val isDoubleClick = Date.now().toLong() - lastMouseUpMillis < 500 && mouseUpCount > 0
            lastMouseUpMillis = Date.now().toLong()
            mouseUpCount += 1

            return isDoubleClick
        }

        fun reset() {
            lastMouseUpMillis = 0L
            mouseUpCount = 0
        }
    }

    companion object {
        private const val SCROLL_SPEED_RATIO = 1 / 1.3F
        private const val SCROLL_THRESHOLD_PIXEL = 1F
    }
}

File: monosketch-svelte/src/lib/mono/workspace/mouse/mouse-event-observer.ts
import { Flow, LifecycleOwner } from '$libs/flow';
import { DrawingInfo } from '$mono/workspace/drawing-info';
import { ScrollMode } from '$mono/ui-state-manager/states';
import { Point, PointF } from '$libs/graphics-geo/point';
import {
    MousePointer,
    MousePointerType,
} from '$mono/workspace/mouse/mouse-pointer';

const SCROLL_SPEED_RATIO = 1 / 1.3;
const SCROLL_THRESHOLD_PIXEL = 1.0;

/**
 * A class to attach and observe mouse events happening on the canvas.
 * All events will be combined into a single flow `mousePointerFlow`.
 */
export class MouseEventObserver {
    private drawingInfo: DrawingInfo;
    private mouseWheelDeltaX: number = 0.0;
    private mouseWheelDeltaY: number = 0.0;

    private drawingOffsetPointPxMutableFlow = new Flow<Point>(Point.ZERO);
    readonly drawingOffsetPointPxFlow = this.drawingOffsetPointPxMutableFlow.immutable();

    private mousePointerMutableFlow = new Flow<MousePointer>(MousePointer.idle);
    readonly mousePointerFlow = this.mousePointerMutableFlow;

    private readonly mouseDoubleClickDetector = new MouseDoubleClickDetector();

    constructor(
        container: HTMLElement,
        private drawingInfoFlow: Flow<DrawingInfo>,
        private scrollModeFlow: Flow<ScrollMode>,
    ) {
        this.drawingInfo = drawingInfoFlow.value!;

        container.onmousedown = this.setMouseDownPointer;
        container.onmouseup = this.setMouseUpPointer;
        container.onmousemove = this.setMouseMovePointer;
        container.onwheel = this.setMouseWheel;
    }

    observeEvents(lifecycleOwner: LifecycleOwner, shiftKeyStateFlow: Flow<boolean>): void {
        this.drawingInfoFlow.observe(lifecycleOwner, (drawingInfo) => {
            this.drawingInfo = drawingInfo;
        });

        shiftKeyStateFlow.observe(lifecycleOwner, (isWithShiftKey) => {
            const currentValue = this.getCurrentMousePointer();
            if (currentValue.type !== MousePointerType.DRAG) {
                return;
            }
            this.mousePointerMutableFlow.value = {
                ...currentValue,
                isWithShiftKey,
            };
        });
    }

    forceUpdateOffset(offsetPx: Point): void {
        this.drawingOffsetPointPxMutableFlow.value = offsetPx;
    }

    private setMouseDownPointer = (event: MouseEvent) => {
        const currentValue = this.getCurrentMousePointer();
        if (
            currentValue.type === MousePointerType.IDLE ||
            currentValue.type === MousePointerType.MOVE
        ) {
            this.mousePointerMutableFlow.value = MousePointer.down(
                this.toBoardCoordinate(event),
                this.toPointPx(event),
                event.shiftKey,
            );
        }
    };

    private setMouseUpPointer = (event: MouseEvent) => {
        const isDoubleClick = this.mouseDoubleClickDetector.onMouseUp();
        const currentValue = this.getCurrentMousePointer();

        const nextValue = () => {
            switch (currentValue.type) {
                case MousePointerType.DOWN:
                    return MousePointer.up(
                        this.toBoardCoordinate(event),
                        this.toBoardCoordinateF(event),
                        currentValue.boardCoordinate,
                        event.shiftKey,
                    );
                case MousePointerType.DRAG:
                    return MousePointer.up(
                        this.toBoardCoordinate(event),
                        this.toBoardCoordinateF(event),
                        currentValue.mouseDownBoardCoordinate,
                        event.shiftKey,
                    );
                default:
                    return MousePointer.idle;
            }
        };
        this.mousePointerMutableFlow.value = nextValue();

        if (currentValue.type === MousePointerType.DOWN) {
            this.mousePointerMutableFlow.value = isDoubleClick
                ? MousePointer.doubleClick(this.toBoardCoordinate(event))
                : MousePointer.click(this.toBoardCoordinate(event), event.shiftKey);
        }

        this.mousePointerMutableFlow.value = MousePointer.idle;
    };

    private setMouseMovePointer = (event: MouseEvent) => {
        this.mouseDoubleClickDetector.reset();

        const currentValue = this.mousePointerMutableFlow.value ?? MousePointer.idle;
        const nextValue = () => {
            switch (currentValue.type) {
                case MousePointerType.DOWN:
                    const mouseBoardCoordinate = this.toBoardCoordinate(event);
                    // If the mouse is moved to a different cell, it's a drag.
                    // otherwise, it remains as a down (no change to the pointer to avoid triggering
                    // new event to listeners).
                    return !mouseBoardCoordinate.equals(currentValue.boardCoordinate)
                        ? MousePointer.drag(
                            this.toBoardCoordinate(event),
                            this.toBoardCoordinateF(event),
                            currentValue.boardCoordinate,
                            event.shiftKey,
                        )
                        : null;
                case MousePointerType.DRAG:
                    return MousePointer.drag(
                        this.toBoardCoordinate(event),
                        this.toBoardCoordinateF(event),
                        currentValue.mouseDownBoardCoordinate,
                        event.shiftKey,
                    );
                case MousePointerType.IDLE:
                case MousePointerType.MOVE:
                case MousePointerType.UP:
                case MousePointerType.CLICK:
                case MousePointerType.DOUBLE_CLICK:
                default:
                    return MousePointer.move(this.toBoardCoordinate(event), this.toPointPx(event));
            }
        };
        const nextMousePointer = nextValue();
        if (nextMousePointer) {
            this.mousePointerMutableFlow.value = nextMousePointer;
        }
    };

    private setMouseWheel = (event: WheelEvent) => {
        event.stopPropagation();
        event.preventDefault();

        const scrollHorizontalThresholdPx: number = SCROLL_THRESHOLD_PIXEL;
        const scrollVerticalThresholdPx: number = SCROLL_THRESHOLD_PIXEL;

        const [scrollDeltaX, scrollDeltaY]: [number, number] = this.getScrollDelta(event);

        const wheelDeltaLeft: number =
            (scrollDeltaX * SCROLL_SPEED_RATIO) / scrollHorizontalThresholdPx;
        const wheelDeltaTop: number =
            (scrollDeltaY * SCROLL_SPEED_RATIO) / scrollVerticalThresholdPx;
        const accumulateWheelDeltaLeft: number = this.mouseWheelDeltaX + wheelDeltaLeft;
        const accumulateWheelDeltaTop: number = this.mouseWheelDeltaY + wheelDeltaTop;

        const usableDeltaLeft: number = Math.floor(accumulateWheelDeltaLeft);
        const usableDeltaTop: number = Math.floor(accumulateWheelDeltaTop);

        if (usableDeltaLeft !== 0 || usableDeltaTop !== 0) {
            const offsetLeft =
                this.drawingInfo.offsetPx.left - usableDeltaLeft * scrollHorizontalThresholdPx;
            const offsetTop =
                this.drawingInfo.offsetPx.top - usableDeltaTop * scrollVerticalThresholdPx;
            this.drawingOffsetPointPxMutableFlow.value = Point.ofF(offsetLeft, offsetTop);
        }

        this.mouseWheelDeltaX = accumulateWheelDeltaLeft - usableDeltaLeft;
        this.mouseWheelDeltaY = accumulateWheelDeltaTop - usableDeltaTop;
    };

    private getCurrentMousePointer = (): MousePointer => this.mousePointerMutableFlow.value!;

    /**
     * Returns scroll delta x and y of the wheel event after adjusting with meta keys:
     * - When Alt/Option key is pressed, swap delta x and delta y (this is for those who use mouse)
     * - When Shift key is pressed, scroll mode other than [ScrollMode.BOTH] will be flipped
     *   (vertical -> horizontal, horizontal -> vertical)
     * - Returns the delta x and delta y of the wheel event according to adjusted scroll mode (with
     *   Shift key) and delta value (with Alt/Option key):
     *   - Vertical: delta x = 0
     *   - Horizontal: delta y = 0
     *   - Both: delta x and delta y = delta value
     */
    private getScrollDelta(wheelEvent: WheelEvent): [number, number] {
        const deltaX: number = wheelEvent.deltaX;
        const deltaY: number = wheelEvent.deltaY;
        const scrollDeltaX: number = wheelEvent.altKey ? deltaY : deltaX;
        const scrollDeltaY: number = wheelEvent.altKey ? deltaX : deltaY;

        const scrollMode: ScrollMode = this.getScrollMode(wheelEvent);

        switch (scrollMode) {
            case ScrollMode.BOTH:
                return [scrollDeltaX, scrollDeltaY];
            case ScrollMode.VERTICAL:
                return [0.0, scrollDeltaY];
            case ScrollMode.HORIZONTAL:
                return [scrollDeltaX, 0.0];
        }
    }

    private getScrollMode = (wheelEvent: WheelEvent): ScrollMode => {
        switch (this.scrollModeFlow.value) {
            case ScrollMode.BOTH:
                return this.scrollModeFlow.value;
            case ScrollMode.VERTICAL:
                return wheelEvent.shiftKey ? ScrollMode.HORIZONTAL : ScrollMode.VERTICAL;
            case ScrollMode.HORIZONTAL:
                return wheelEvent.shiftKey ? ScrollMode.VERTICAL : ScrollMode.HORIZONTAL;
            default:
                throw Error(`Unknown scroll mode: ${this.scrollModeFlow.value}`);
        }
    };

    private toBoardCoordinate = (mouseEvent: MouseEvent): Point =>
        new Point(
            this.drawingInfo.toBoardColumn(mouseEvent.offsetX),
            this.drawingInfo.toBoardRow(mouseEvent.offsetY),
        );

    private toBoardCoordinateF = (mouseEvent: MouseEvent): PointF =>
        new PointF(
            this.drawingInfo.toBoardColumnF(mouseEvent.offsetX),
            this.drawingInfo.toBoardRowF(mouseEvent.offsetY),
        );

    private toPointPx = (mouseEvent: MouseEvent): Point =>
        new Point(mouseEvent.offsetX, mouseEvent.offsetY);
}

/**
 * A class for detecting double click.
 * If two mouse ups happen within 500ms, it's a double click.
 */
class MouseDoubleClickDetector {
    private lastMouseUpMillis: number = 0;
    private mouseUpCount: number = 0;

    onMouseUp(): boolean {
        const currentTime = Date.now();
        const isDoubleClick = currentTime - this.lastMouseUpMillis < 500 && this.mouseUpCount > 0;
        this.lastMouseUpMillis = currentTime;
        this.mouseUpCount += 1;

        return isDoubleClick;
    }

    reset(): void {
        this.lastMouseUpMillis = 0;
        this.mouseUpCount = 0;
    }
}
